File: rawDroid/src/main/java/com/anthonymandra/rawdroid/XmpEditFragment.java
package com.anthonymandra.rawdroid;

import android.os.Bundle;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;

import com.anthonymandra.widget.XmpLabelGroup;

import java.util.Collection;
import java.util.List;

import uk.co.deanwild.materialshowcaseview.MaterialShowcaseSequence;
import uk.co.deanwild.materialshowcaseview.MaterialShowcaseView;

public class XmpEditFragment extends XmpBaseFragment
{
	private static final XmpEditValues recentXmp = new XmpEditValues();

	public interface RatingChangedListener
	{
		void onRatingChanged(Integer rating);
	}

	public interface LabelChangedListener
	{
		void onLabelChanged(String label);
	}

	public interface SubjectChangedListener
	{
		void onSubjectChanged(String[] subject);
	}

	private RatingChangedListener mRatingListener;
	private LabelChangedListener mLabelListener;
	private SubjectChangedListener mSubjectListener;

	public void setRatingListener(RatingChangedListener listener)
	{
		mRatingListener = listener;
	}

	public void setLabelListener(LabelChangedListener listener)
	{
		mLabelListener = listener;
	}

	public void setSubjectListener(SubjectChangedListener listener)
	{
		mSubjectListener = listener;
	}

	private MetaChangedListener mXmpChangedListener;
	public interface MetaChangedListener
	{
		void onMetaChanged(Integer rating, String label, String[] subject);
	}

	public void setListener(MetaChangedListener listener)
	{
		mXmpChangedListener = listener;
	}

	@Override
	protected void onXmpChanged(XmpValues xmp)
	{
		recentXmp.Subject = xmp.subject;
		recentXmp.Rating = formatRating(xmp.rating);
		recentXmp.Label = formatLabel(xmp.label);

		fireMetaUpdate();
	}

	private void fireMetaUpdate()
	{
		mXmpChangedListener.onMetaChanged(
				recentXmp.Rating,
				recentXmp.Label,
				recentXmp.Subject);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		return inflater.inflate(R.layout.xmp_edit_landscape, container, false);
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState)
	{
		super.onActivityCreated(savedInstanceState);
		setMultiselect(false);
		attachButtons();
	}

	/**
	 * Convenience method for single select XmpLabelGroup
	 */
	private String getLabel()
	{
		String[] label = getColorLabels();
		return label != null ? label[0] : null;
	}

	/**
	 * Convenience method for single select XmpLabelGroup
	 */
	private Integer getRating()
	{
		Integer[] ratings = getRatings();
		return ratings != null ? ratings[0] : null;
	}

	protected Integer formatRating(Integer[] ratings)
	{
		if (ratings == null)
			return null;
		else
			return ratings[0];
	}

	protected Integer[] formatRating(Integer rating)
	{
		if (rating == null)
			return null;
		else
			return new Integer[]{rating};
	}

	protected String formatLabel(String[] labels)
	{
		if (labels == null)
			return null;
		else
			return labels[0];
	}

	protected String[] formatLabel(String label)
	{
		if (label == null)
			return null;
		else
			return new String[]{label};
	}

	/**
	 * Silently set xmp without firing listeners
	 */
	public void initXmp(Integer rating, String[] subject, String label)
	{
		super.initXmp(formatRating(rating),
				formatLabel(label),
				subject);
	}

	private void attachButtons()
	{
		getView().findViewById(R.id.clearMetaButton).setOnClickListener(new OnClickListener()
		{
			@Override
			public void onClick(View v)
			{
				clear();
			}
		});
		getView().findViewById(R.id.recentMetaButton).setOnClickListener(new OnClickListener()
		{
			@Override
			public void onClick(View v)
			{
				fireMetaUpdate();
			}
		});

		final ImageButton helpButton = (ImageButton) getView().findViewById(R.id.helpButton);
		helpButton.setOnClickListener(new View.OnClickListener()
		{
			@Override
			public void onClick(View v)
			{
				startTutorial();
			}
		});
	}

	@Override
	public void onKeywordsSelected(Collection<String> selectedKeywords)
	{
		recentXmp.Subject = getSubject();
		if (mSubjectListener != null)
			mSubjectListener.onSubjectChanged(recentXmp.Subject);
	}

	@Override
	public void onLabelSelectionChanged(List<XmpLabelGroup.Labels> checked)
	{
		if (checked.size() > 0)
		{
			switch (checked.get(0))
			{
				case blue:
					((ImageView)getView().findViewById(R.id.recentLabel)).setImageTintList(ContextCompat.getColorStateList(getContext(), R.color.colorKeyBlue));
					break;
				case red:
					((ImageView)getView().findViewById(R.id.recentLabel)).setImageTintList(ContextCompat.getColorStateList(getContext(), R.color.colorKeyRed));
					break;
				case green:
					((ImageView)getView().findViewById(R.id.recentLabel)).setImageTintList(ContextCompat.getColorStateList(getContext(), R.color.colorKeyGreen));
					break;
				case yellow:
					((ImageView)getView().findViewById(R.id.recentLabel)).setImageTintList(ContextCompat.getColorStateList(getContext(), R.color.colorKeyYellow));
					break;
				case purple:
					((ImageView)getView().findViewById(R.id.recentLabel)).setImageTintList(ContextCompat.getColorStateList(getContext(), R.color.colorKeyPurple));
					break;
			}
		}
		else
		{
			((ImageView)getView().findViewById(R.id.recentLabel)).setImageTintList(ContextCompat.getColorStateList(getContext(), R.color.white));
		}
		recentXmp.Label = getLabel();
		if (mLabelListener != null)
			mLabelListener.onLabelChanged(recentXmp.Label);
	}

	@Override
	public void onRatingSelectionChanged(List<Integer> checked)
	{
		recentXmp.Rating = getRating();
		if (mRatingListener != null)
			mRatingListener.onRatingChanged(recentXmp.Rating);

		if (recentXmp.Rating == null)
		{
			((ImageView) getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star_border);
			return;
		}

		switch (recentXmp.Rating)
		{
			case 5:
				((ImageView)getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star5);
				break;
			case 4:
				((ImageView)getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star4);
				break;
			case 3:
				((ImageView)getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star3);
				break;
			case 2:
				((ImageView)getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star2);
				break;
			case 1:
				((ImageView)getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star1);
				break;
			default:
				((ImageView)getView().findViewById(R.id.recentRating)).setImageResource(R.drawable.ic_star_border);
				break;
		}
	}

	/**
	 * Default values indicate no xmp
	 */
	public static class XmpEditValues
	{
		public Integer Rating = null;
		public String[] Subject = null;
		public String Label = null;
	}

	private void startTutorial()
	{
		MaterialShowcaseSequence sequence = new MaterialShowcaseSequence(getActivity());

		// Sort group
		View sortgroup = getView().findViewById(R.id.recentMetaButton);
		sequence.addSequenceItem(getRectangularView(
				sortgroup,
				R.string.tutSetRecent));

		// Segregate
		View segregate = getView().findViewById(R.id.clearMetaButton);
		sequence.addSequenceItem(getRectangularView(
				segregate,
				R.string.tutClearMeta));

		// Rating
		View labelRating = getView().findViewById(R.id.metaLabelRating);
		sequence.addSequenceItem(getRectangularView(
				labelRating,
				R.string.tutSetRatingLabel));

		// Subject
		View subject = getView().findViewById(R.id.keywordFragment);
		sequence.addSequenceItem(getRectangularView(
				subject,
				R.string.tutSetSubject));

		// Match
		View addKeyword = getView().findViewById(R.id.addKeyword);
		sequence.addSequenceItem(getRectangularView(
				addKeyword,
				R.string.tutAddSubject));

		// Match
		View editKeyword = getView().findViewById(R.id.editKeyword);
		sequence.addSequenceItem(getRectangularView(
				editKeyword,
				R.string.tutEditSubject));

		sequence.start();
	}

	private MaterialShowcaseView getRectangularView(View target, @StringRes int contentId)
	{
		return getRectangularView(target,
				R.string.editMetadata,
				contentId,
				R.string.ok);
	}

	private MaterialShowcaseView getRectangularView(View target, @StringRes int titleId, @StringRes int contentId, @StringRes int dismissId)
	{
		return getRectangularView(target,
				getString(titleId),
				getString(contentId),
				getString(dismissId));
	}

	private MaterialShowcaseView getRectangularView(View target, String title, String content, String dismiss)
	{
		return new MaterialShowcaseView.Builder(getActivity())
				.setTarget(target)
				.setTitleText(title)
				.setContentText(content)
				.setDismissOnTouch(true)
				.setDismissText(dismiss)
				.withRectangleShape()
				.build();
	}
}

File: app/src/main/java/com/anthonymandra/curator/XmpEditFragment.kt
package com.anthonymandra.curator

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import com.anthonymandra.curator.databinding.XmpEditLandscapeBinding
import com.anthonymandra.curator.databinding.XmpSubjectEditBinding
import com.anthonymandra.curator.data.Label
import com.anthonymandra.curator.data.SubjectEntity
import uk.co.deanwild.materialshowcaseview.MaterialShowcaseSequence
import uk.co.deanwild.materialshowcaseview.MaterialShowcaseView
import java.util.*

typealias RatingChangedListener = (rating: Int?) -> Unit
typealias LabelChangedListener = (label: String?) -> Unit
typealias SubjectChangedListener = (subject: Collection<SubjectEntity>?) -> Unit
typealias MetaChangedListener = (xmp: XmpValues) -> Unit
class XmpEditFragment : XmpBaseFragment() {
    private var _ui: XmpEditLandscapeBinding? = null
    private lateinit var uiSubject: XmpSubjectEditBinding


    // This property is only valid between onCreateView and onDestroyView.
    private val ui get() = _ui!!

    private var mRatingListener: RatingChangedListener? = null
    private var mLabelListener: LabelChangedListener? = null
    private var mSubjectListener: SubjectChangedListener? = null

    private var mXmpChangedListener: MetaChangedListener? = null

    /**
     * Convenience method for single select XmpLabelGroup
     */
    private val label: String?
			get() { return colorLabels.firstOrNull() }

    /**
     * Convenience method for single select XmpLabelGroup
     */
    private val rating: Int?
			get() { return ratings.firstOrNull() }

    fun setRatingListener(listener: RatingChangedListener) {
        mRatingListener = listener
    }

    fun setLabelListener(listener: LabelChangedListener) {
        mLabelListener = listener
    }

    fun setSubjectListener(listener: SubjectChangedListener) {
        mSubjectListener = listener
    }

    fun setListener(listener: MetaChangedListener) {
        mXmpChangedListener = listener
    }

    override fun onXmpChanged(xmp: XmpFilter) {
        recentXmp.rating = xmp.rating.firstOrNull()
        recentXmp.label = xmp.label.firstOrNull()
        recentXmp.subject = xmp.subject

				mXmpChangedListener?.invoke(recentXmp)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _ui = XmpEditLandscapeBinding.inflate(inflater, container, false)
        return ui.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _ui = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isSingleSelection(true)
//        setAllowUnselected(true)

        ui.clearMetaButton.setOnClickListener { clear() }
        ui.recentMetaButton.setOnClickListener { mXmpChangedListener?.invoke(recentXmp) }
        ui.helpButton.setOnClickListener { startTutorial() }
    }

    /**
     * Silently set xmp without firing listeners
     */
    fun initXmp(rating: Int?, subject: List<SubjectEntity>, label: String?) {
        super.initXmp(
						if (rating == null) Collections.emptyList() else arrayListOf(rating),
						if (label == null) Collections.emptyList() else arrayListOf(label),
                subject)
    }

    override fun onKeywordsSelected(selectedKeywords: Collection<SubjectEntity>) {
        recentXmp.subject = subject
        mSubjectListener?.invoke(recentXmp.subject)
    }

    override fun onLabelSelectionChanged(checked: List<Label>) {
        val c = context
        if (c != null) {
            if (checked.isNotEmpty()) {
                when (checked[0]) {
                    Label.Blue -> ui.recentLabel.imageTintList = ContextCompat.getColorStateList(c, R.color.colorKeyBlue)
                    Label.Red -> ui.recentLabel.imageTintList = ContextCompat.getColorStateList(c, R.color.colorKeyRed)
                    Label.Green -> ui.recentLabel.imageTintList = ContextCompat.getColorStateList(c, R.color.colorKeyGreen)
                    Label.Yellow -> ui.recentLabel.imageTintList = ContextCompat.getColorStateList(c, R.color.colorKeyYellow)
                    Label.Purple -> ui.recentLabel.imageTintList = ContextCompat.getColorStateList(c, R.color.colorKeyPurple)
                }
            } else {
                ui.recentLabel.imageTintList = ContextCompat.getColorStateList(c, R.color.white)
            }
        }

        recentXmp.label = label
        mLabelListener?.invoke(recentXmp.label)
    }

    override fun onRatingSelectionChanged(checked: List<Int>) {
        recentXmp.rating = rating
        mRatingListener?.invoke(recentXmp.rating)

        if (recentXmp.rating == null) {
            ui.recentRating.setImageResource(R.drawable.ic_star_border)
            return
        }

        when (recentXmp.rating) {
            5 -> ui.recentRating.setImageResource(R.drawable.ic_star5)
            4 -> ui.recentRating.setImageResource(R.drawable.ic_star4)
            3 -> ui.recentRating.setImageResource(R.drawable.ic_star3)
            2 -> ui.recentRating.setImageResource(R.drawable.ic_star2)
            1 -> ui.recentRating.setImageResource(R.drawable.ic_star1)
            else -> ui.recentRating.setImageResource(R.drawable.ic_star_border)
        }
    }

    private fun startTutorial() {
        val sequence = MaterialShowcaseSequence(activity)

        // Sort group
        sequence.addSequenceItem(getRectangularView(
            ui.recentMetaButton,
                R.string.tutSetRecent))

        // Segregate
        sequence.addSequenceItem(getRectangularView(
            ui.clearMetaButton,
                R.string.tutClearMeta))

        // rating
        sequence.addSequenceItem(getRectangularView(
            ui.metaLabelRating.root,
                R.string.tutSetRatingLabel))

        // subject
        sequence.addSequenceItem(getRectangularView(
            uiSubject.root,
            R.string.tutSetSubject))

        // Match
        sequence.addSequenceItem(getRectangularView(
            uiSubject.addKeyword,
                R.string.tutAddSubject))

        // Match
        sequence.addSequenceItem(getRectangularView(
            uiSubject.editKeyword,
                R.string.tutEditSubject))

        sequence.start()
    }

    private fun getRectangularView(target: View, @StringRes contentId: Int): MaterialShowcaseView {
        return getRectangularView(target,
                getString(R.string.editMetadata),
                getString(contentId),
                getString(R.string.ok))
    }

    private fun getRectangularView(target: View, title: String, content: String, dismiss: String): MaterialShowcaseView {
        return MaterialShowcaseView.Builder(activity)
                .setTarget(target)
                .setTitleText(title)
                .setContentText(content)
                .setDismissOnTouch(true)
                .setDismissText(dismiss)
                .withRectangleShape()
                .build()
    }

    companion object {
        private val recentXmp = XmpValues()
    }
}
